ui/src/components/pages/Pipeline/partials/Graph/index.tsx

```diff
@@ -74,6 +85,101 @@ const getLayoutedElements = (
   return { nodes, edges };
 };
 
+const Flow = (props: FlowProps) => {
+  const zoomLevel = useViewport().zoom;
+  const { zoomIn, zoomOut, fitView } = useReactFlow();
+  const [isLocked, setIsLocked] = useState(false);
+  const [isScrollLocked, setIsScrollLocked] = useState(false);
+  const {
+    nodes,
+    edges,
+    onNodesChange,
+    onEdgesChange,
+    onConnect,
+    handleNodeClick,
+    handleEdgeClick,
+    handlePaneClick,
+  } = props;
+
+  return (
+    <ReactFlow
+      nodeTypes={defaultNodeTypes}
+      edgeTypes={defaultEdgeTypes}
+      nodes={nodes}
+      edges={edges}
+      onNodesChange={onNodesChange}
+      onEdgesChange={onEdgesChange}
+      onConnect={onConnect}
+      onEdgeClick={handleEdgeClick}
+      onNodeClick={handleNodeClick}
+      onPaneClick={handlePaneClick}
+      fitView
+      preventScrolling={isScrollLocked}
+      panOnDrag={!isLocked}
+    >
+      <Panel position="bottom-left">
+        <IconButton onClick={() => setIsLocked((prevState) => !prevState)}>
+          <img src={isLocked ? lock : unlock} alt={"lock"} />
+        </IconButton>
+        <IconButton
+          onClick={() => setIsScrollLocked((prevState) => !prevState)}
+        >
+          <img src={scrollToggle} alt={"scrollLock"} />
+        </IconButton>
+        <div className={"divider"} />
+        <IconButton onClick={() => fitView()}>
+          <img src={fullscreen} alt={"fullscreen"} />
+        </IconButton>
+        <IconButton
+          onClick={() => {
```

Comment:
Would be good to move all these in-line callback method definitions to useCallback methods. It will result in less overall re-rendering of children.

Reply:
sure, let me do that